CMakeLists.txt updates for modern CMake

nimish opened the pull request:
Update cmake to 3.21+ and simplify the logic

Uses the bundled FindODBC and uses dependent options

 - [x] Review
 - [x] Adjust for comments
 - [x] All CI builds and checks have passed

mloskot commented:
This is great, thank you.

First, I'd like to hear from @jimhester @detule @lexicalunit  if they have any objections to bump the required version of CMake.

nimish commented:
> This is great, thank you.
> 
> First, I'd like to hear from @jimhester @detule @lexicalunit if they have any objections to bump the required version of CMake.

Sure. I think 3.21 was conservative to get all the relevant policies. It could probably be relaxed.

detule commented:
Hi there: took a quick look.

I probably read the diff wrong, but does this also bump the required language standard to c++17? 

Also the comment
```
# #######################################
# # require and enable C++20
# #######################################
```
seems misleading.

nimish commented:
> Hi there: took a quick look.
> 
> I probably read the diff wrong, but does this also bump the required language standard to c++17?
> 
> Also the comment
> 
> ```
> # #######################################
> # # require and enable C++20
> # #######################################
> ```
> 
> seems misleading.

Oops, that was from a different branch that shouldn't have made it in here. Relaxed back to 14.

mloskot commented:
Good catch @detule!

@nimish  Yes, we don't want to switch away from C++14 yet, especially there are important users like r-odbc which may still need C++14. If we switch, it will be C++17 though, see https://github.com/nanodbc/nanodbc/discussions/321 However, we can accept conditionally-compiled support for newer C++ standards.

nimish commented:
> Good catch @detule!
> 
> @nimish Yes, we don't want to switch away from C++14 yet, especially there are important users like r-odbc which may still need C++14. If we switch, it will be C++17 though, see #321 However, we can accept conditionally-compiled support for newer C++ standards.

No problem, was testing out C++20 support in general. 

Can't quite understand the reason the CI is now failing

mloskot commented:
> Can't quite understand the reason the CI is now failing

Sorry, I didn't have time to check. I will try to look at the CI results tomorrow and see if I can help

mloskot commented on .github/workflows/ci-linux.yml:
```diff
@@ -56,7 +56,7 @@ jobs:
       - name: Install
         run: |
-          sudo apt-get install -y cmake ${{ matrix.cxx }}
+          sudo apt-get install -y cmake libboost-locale-dev libboost-dev ${{ matrix.cxx }}
```
Boost is only required if CMake configuration phase is specified with
https://github.com/nanodbc/nanodbc/blob/5b52ff0c8fd70572ed610bab8104fd5be08d954e/CMakeLists.txt#L19

nimish replied:
Yep, looks like I defaulted to using Boost.Locale if Unicode support was built in. Fixed

mloskot commented on example/empty.cpp:
```diff
@@ -1,4 +1,3 @@
-#include "example_unicode_utils.h"
```
Was the file renaming intentional here?

nimish replied:
Yes -- the cmake `configure_file` didn't interpolate anything as far as I could tell

mloskot replied:
Yes, there is nothing to interpolate indeed.

The unobvious reason why it there is `empty.in.cpp` in the repo and not `empty.cpp` is to allow this workflow
- CMake creates `empty.cpp` 
- The `empty.cpp` is available to contributors and users as playground for experiments, to modify, build and debug as part of the nanodbc project.
- The `empty.cpp` is automatically ignored, so contributors don't have to be cautious to not to commit their experiments to the nanodbc repo or forks

mloskot replied:
Thanks @nimish 
If there are better ways to achieve similar, then please feel free to propose a change.

nimish replied:
> Thanks @nimish
> If there are better ways to achieve similar, then please feel free to propose a change.

Honestly, not sure -- this does seem like a decent solution. It's documented in a comment now too

nimish commented:
@mloskot 
* reverted empty.cpp changes
* fixed some cmake defaulting issues with vars
* added some inline comments to help explain some of the non-obvious bits

Workflows will still need approval, but it should be back to all the same defaults as stock nanodbc. Might be interesting to also add boost to the ci matrix.

nimish commented on .gitignore:
```diff
@@ -33,6 +33,7 @@ package-lock.json
 *.code-workspace
 
 # Miscelenaous
+.cache
```
for `clangd` analysis, it creates a `.cache` folder

mloskot commented:
> Workflows will still need approval, but it should be back to all the same defaults as stock nanodbc. 

There is one failure for MySQL test due to SSL issue which I'm not clear about, but I think this can be ignored for this PR.

> Might be interesting to also add boost to the ci matrix.

Yes, indeed.

We switched to GHA from Travis CI and AppVeyor quite recently, so the CI coverage is still a work in progress (see #265).

> Also, the fallbacks for ancient __GNUC__ could be removed given the minimum req of C++14

Good catch! Opened #383

I'm happy to merge this PR. Thank you for your contribution.

mloskot commented:
> Do we need the appveyor env var check?

No, we no longer build on AV, last time was two months ago https://ci.appveyor.com/project/nanodbc/nanodbc
Another good catch to clean up.

nimish commented:
Cool, there's some minor typos to clean up as well. https://dev.mysql.com/doc/refman/8.0/en/connection-options.html#option_general_ssl-mode says `ssl-mode=PREFERRED` is the default and falls back to unencrypted, but this run just breaks anyway
